File: src/modules/equipment-assignment/services/equipment-assignment.service.ts
import {
  BadRequestException,
  ForbiddenException,
  Injectable,
  NotFoundException,
} from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { AccessControlService } from '../../../common/services/access-control.service';
import { AuthenticatedUser } from '../../../common/types/authenticated-user.type';
import { Employee } from '../../employee/entities/employee.entity';
import { EquipmentAsset } from '../../equipment-asset/entities/equipment-asset.entity';
import { EquipmentAssetStatus } from '../../equipment-asset/enums/equipment-asset-status.enum';
import { NotificationService } from '../../notification/services/notification.service';
import { ReturnRequestDto } from '../dto/return-request.dto';
import { EquipmentAssignment } from '../entities/equipment-assignment.entity';
import { EquipmentAssignmentStatus } from '../enums/equipment-assignment-status.enum';

@Injectable()
export class EquipmentAssignmentService {
  constructor(
    @InjectRepository(EquipmentAssignment)
    private readonly assignmentRepository: Repository<EquipmentAssignment>,
    @InjectRepository(EquipmentAsset)
    private readonly assetRepository: Repository<EquipmentAsset>,
    @InjectRepository(Employee)
    private readonly employeeRepository: Repository<Employee>,
    private readonly notificationService: NotificationService,
    private readonly accessControl: AccessControlService,
  ) {}

  findMyAssignments(userId: string): Promise<EquipmentAssignment[]> {
    return this.assignmentRepository.find({
      where: { employee: { id: userId } },
      relations: {
        equipmentAsset: { equipmentModel: { category: true } },
        employee: true,
        assignedBy: true,
        returnRequestedBy: true,
      },
      order: { assignedAt: 'DESC' },
    });
  }

  async findOne(
    id: string,
    user: AuthenticatedUser,
  ): Promise<EquipmentAssignment> {
    const assignment = await this.assignmentRepository.findOne({
      where: { id },
      relations: {
        equipmentAsset: { equipmentModel: { category: true } },
        employee: { department: true },
        assignedBy: true,
        returnRequestedBy: true,
        request: true,
      },
    });

    if (!assignment) {
      throw new NotFoundException(`Assignment with id "${id}" not found`);
    }

    await this.assertCanViewAssignment(assignment, user);
    return assignment;
  }

  async findTeamEquipment(
    user: AuthenticatedUser,
  ): Promise<EquipmentAssignment[]> {
    const departmentIds = await this.accessControl.getManagedDepartmentIds(
      user.id,
    );
    if (!departmentIds.length) {
      return [];
    }

    const employeeIds =
      await this.accessControl.getDepartmentEmployeeIds(departmentIds);

    return this.assignmentRepository
      .createQueryBuilder('assignment')
      .leftJoinAndSelect('assignment.equipmentAsset', 'asset')
      .leftJoinAndSelect('asset.equipmentModel', 'model')
      .leftJoinAndSelect('model.category', 'category')
      .leftJoinAndSelect('assignment.employee', 'employee')
      .where('employee.id IN (:...employeeIds)', { employeeIds })
      .andWhere('assignment.status IN (:...statuses)', {
        statuses: [
          EquipmentAssignmentStatus.ACTIVE,
          EquipmentAssignmentStatus.RETURN_REQUESTED,
        ],
      })
      .orderBy('assignment.assignedAt', 'DESC')
      .getMany();
  }

  async requestReturn(
    id: string,
    user: AuthenticatedUser,
    dto: ReturnRequestDto,
  ): Promise<EquipmentAssignment> {
    const assignment = await this.findOne(id, user);

    if (!this.accessControl.isDirectManagerOrAbove(user)) {
      throw new ForbiddenException(
        'Only managers can request equipment returns',
      );
    }

    const canManage = await this.accessControl.isEmployeeInManagedDepartments(
      user.id,
      assignment.employee.id,
    );
    if (!canManage && !this.accessControl.isProcurementManagerOrAbove(user)) {
      throw new ForbiddenException(
        'You cannot request return for this assignment',
      );
    }

    if (assignment.status !== EquipmentAssignmentStatus.ACTIVE) {
      throw new BadRequestException(
        'Only active assignments can have return requested',
      );
    }

    const requester = await this.employeeRepository.findOne({
      where: { id: user.id },
    });

    assignment.status = EquipmentAssignmentStatus.RETURN_REQUESTED;
    assignment.returnRequestedBy = requester ?? undefined;
    assignment.returnRequestedAt = new Date();
    assignment.returnByDate = dto.returnByDate;
    assignment.returnNote = dto.message;

    const asset = assignment.equipmentAsset;
    asset.status = EquipmentAssetStatus.RETURN_REQUESTED;
    await this.assetRepository.save(asset);

    const saved = await this.assignmentRepository.save(assignment);
    await this.notificationService.notifyEquipmentReturnRequested(
      assignment.employee,
      saved,
      dto.returnByDate,
      dto.message,
    );

    return saved;
  }

  async completeReturn(
    id: string,
    user: AuthenticatedUser,
  ): Promise<EquipmentAssignment> {
    const assignment = await this.findOne(id, user);

    if (
      assignment.employee.id !== user.id &&
      !this.accessControl.isDirectManagerOrAbove(user)
    ) {
      throw new ForbiddenException('You cannot complete this return');
    }

    if (assignment.status !== EquipmentAssignmentStatus.RETURN_REQUESTED) {
      throw new BadRequestException('Assignment is not awaiting return');
    }

    assignment.status = EquipmentAssignmentStatus.RETURNED;
    assignment.returnedAt = new Date();

    const asset = assignment.equipmentAsset;
    asset.status = EquipmentAssetStatus.AVAILABLE;
    asset.assignedEmployee = undefined;
    asset.assignedAt = undefined;
    asset.expectedReturnDate = undefined;
    await this.assetRepository.save(asset);

    const saved = await this.assignmentRepository.save(assignment);
    await this.notificationService.notifyEquipmentReturned(
      assignment.employee,
      saved,
    );

    return saved;
  }

  private async assertCanViewAssignment(
    assignment: EquipmentAssignment,
    user: AuthenticatedUser,
  ): Promise<void> {
    if (assignment.employee.id === user.id) return;
    if (this.accessControl.isProcurementManagerOrAbove(user)) return;

    if (this.accessControl.isDirectManagerOrAbove(user)) {
      const canView = await this.accessControl.isEmployeeInManagedDepartments(
        user.id,
        assignment.employee.id,
      );
      if (canView) return;
    }

    throw new ForbiddenException('You cannot view this assignment');
  }
}

File: src/EquipmentApi/Services/EquipmentAssignmentService.cs
using EquipmentApi.Auth;
using EquipmentApi.Data;
using EquipmentApi.Dtos;
using EquipmentApi.Entities;
using EquipmentApi.Enums;
using EquipmentApi.Exceptions;
using Microsoft.EntityFrameworkCore;

namespace EquipmentApi.Services;

public class EquipmentAssignmentService
{
    private readonly AppDbContext _db;
    private readonly NotificationService _notificationService;
    private readonly AccessControlService _accessControl;

    public EquipmentAssignmentService(
        AppDbContext db,
        NotificationService notificationService,
        AccessControlService accessControl)
    {
        _db = db;
        _notificationService = notificationService;
        _accessControl = accessControl;
    }

    public Task<List<EquipmentAssignment>> FindMyAssignmentsAsync(
        Guid userId,
        CancellationToken cancellationToken = default)
    {
        return _db.EquipmentAssignments
            .Where(assignment => assignment.EmployeeId == userId)
            .Include(assignment => assignment.EquipmentAsset)
                .ThenInclude(asset => asset.EquipmentModel)
                    .ThenInclude(model => model.Category)
            .Include(assignment => assignment.Employee)
            .Include(assignment => assignment.AssignedBy)
            .Include(assignment => assignment.ReturnRequestedBy)
            .OrderByDescending(assignment => assignment.AssignedAt)
            .ToListAsync(cancellationToken);
    }

    public async Task<EquipmentAssignment> FindOneAsync(
        Guid id,
        AuthenticatedUser user,
        CancellationToken cancellationToken = default)
    {
        var assignment = await LoadAssignmentAsync(id, cancellationToken);
        if (assignment is null)
        {
            throw new NotFoundException($"Assignment with id \"{id}\" not found");
        }

        await AssertCanViewAssignmentAsync(assignment, user, cancellationToken);
        return assignment;
    }

    public async Task<List<EquipmentAssignment>> FindTeamEquipmentAsync(
        AuthenticatedUser user,
        CancellationToken cancellationToken = default)
    {
        var departmentIds = await _accessControl.GetManagedDepartmentIdsAsync(user.Id, cancellationToken);
        if (departmentIds.Count == 0)
        {
            return [];
        }

        var employeeIds = await _accessControl.GetDepartmentEmployeeIdsAsync(departmentIds, cancellationToken);
        if (employeeIds.Count == 0)
        {
            return [];
        }

        return await _db.EquipmentAssignments
            .Where(assignment => employeeIds.Contains(assignment.EmployeeId)
                && (assignment.Status == EquipmentAssignmentStatus.Active
                    || assignment.Status == EquipmentAssignmentStatus.ReturnRequested))
            .Include(assignment => assignment.EquipmentAsset)
                .ThenInclude(asset => asset.EquipmentModel)
                    .ThenInclude(model => model.Category)
            .Include(assignment => assignment.Employee)
            .OrderByDescending(assignment => assignment.AssignedAt)
            .ToListAsync(cancellationToken);
    }

    public async Task<EquipmentAssignment> RequestReturnAsync(
        Guid id,
        AuthenticatedUser user,
        ReturnRequestDto dto,
        CancellationToken cancellationToken = default)
    {
        var assignment = await FindOneAsync(id, user, cancellationToken);

        if (!_accessControl.IsDirectManagerOrAbove(user))
        {
            throw new ForbiddenException("Only managers can request equipment returns");
        }

        var canManage = await _accessControl.IsEmployeeInManagedDepartmentsAsync(
            user.Id,
            assignment.EmployeeId,
            cancellationToken);

        if (!canManage && !_accessControl.IsProcurementManagerOrAbove(user))
        {
            throw new ForbiddenException("You cannot request return for this assignment");
        }

        if (assignment.Status != EquipmentAssignmentStatus.Active)
        {
            throw new BadRequestException("Only active assignments can have return requested");
        }

        assignment.Status = EquipmentAssignmentStatus.ReturnRequested;
        assignment.ReturnRequestedById = user.Id;
        assignment.ReturnRequestedAt = DateTime.UtcNow;
        assignment.ReturnByDate = DateOnly.Parse(dto.ReturnByDate);
        assignment.ReturnNote = dto.Message;

        assignment.EquipmentAsset.Status = EquipmentAssetStatus.ReturnRequested;
        await _db.SaveChangesAsync(cancellationToken);

        await _notificationService.NotifyEquipmentReturnRequestedAsync(
            assignment.Employee,
            assignment,
            dto.ReturnByDate,
            dto.Message,
            cancellationToken);

        return assignment;
    }

    public async Task<EquipmentAssignment> CompleteReturnAsync(
        Guid id,
        AuthenticatedUser user,
        CancellationToken cancellationToken = default)
    {
        var assignment = await FindOneAsync(id, user, cancellationToken);

        if (assignment.EmployeeId != user.Id && !_accessControl.IsDirectManagerOrAbove(user))
        {
            throw new ForbiddenException("You cannot complete this return");
        }

        if (assignment.Status != EquipmentAssignmentStatus.ReturnRequested)
        {
            throw new BadRequestException("Assignment is not awaiting return");
        }

        assignment.Status = EquipmentAssignmentStatus.Returned;
        assignment.ReturnedAt = DateTime.UtcNow;

        assignment.EquipmentAsset.Status = EquipmentAssetStatus.Available;
        assignment.EquipmentAsset.AssignedEmployeeId = null;
        assignment.EquipmentAsset.AssignedAt = null;
        assignment.EquipmentAsset.ExpectedReturnDate = null;

        await _db.SaveChangesAsync(cancellationToken);
        await _notificationService.NotifyEquipmentReturnedAsync(
            assignment.Employee,
            assignment,
            cancellationToken);

        return assignment;
    }

    private async Task AssertCanViewAssignmentAsync(
        EquipmentAssignment assignment,
        AuthenticatedUser user,
        CancellationToken cancellationToken)
    {
        if (assignment.EmployeeId == user.Id)
        {
            return;
        }

        if (_accessControl.IsProcurementManagerOrAbove(user))
        {
            return;
        }

        if (_accessControl.IsDirectManagerOrAbove(user))
        {
            var canView = await _accessControl.IsEmployeeInManagedDepartmentsAsync(
                user.Id,
                assignment.EmployeeId,
                cancellationToken);
            if (canView)
            {
                return;
            }
        }

        throw new ForbiddenException("You cannot view this assignment");
    }

    private Task<EquipmentAssignment?> LoadAssignmentAsync(Guid id, CancellationToken cancellationToken)
    {
        return _db.EquipmentAssignments
            .Include(assignment => assignment.EquipmentAsset)
                .ThenInclude(asset => asset.EquipmentModel)
                    .ThenInclude(model => model.Category)
            .Include(assignment => assignment.Employee)
                .ThenInclude(employee => employee.Department)
            .Include(assignment => assignment.AssignedBy)
            .Include(assignment => assignment.ReturnRequestedBy)
            .Include(assignment => assignment.Request)
            .FirstOrDefaultAsync(assignment => assignment.Id == id, cancellationToken);
    }
}
